README.md
# PC Importer

**IMPORTANT** 
* For Foundry 0.8.x and 9.x use [PC Importer v0.8.x](https://github.com/spetzel2020/pc-importer/releases/latest/download/module.json)
* For Foundry 0.7.x and earlier use [PC Importer v0.6.8](https://raw.githubusercontent.com/spetzel2020/pc-importer/v0.6.8/module.json)

**Author**: Spetzel#0103
**Version**: 0.8.0
**Foundry VTT Compatibility**: 0.8.x and 9 (for Foundry 0.7.x and earlier, use the link above)
**System Compatibility (If applicable)**: dnd5e
**Translation Support**: en


# Description
PC Importer imports Player Character sheets from other systems and methods into Foundry VTT Actors, matching to SRD or other Compendiums you have created manually or through importers such as D&D Beyond.

# Install
1. Go to the "Add-on Modules" tab in Foundry Setup
2. Click "Install Module" and paste this link in the field at the bottom: `https://github.com/spetzel2020/pc-importer/releases/latest/download/module.json`
3. Open your world and go to Settings>Manage Modules and enable **PC Importer**

## Using PC Importer with MPMB sheets
### With your MPMB character sheet
1. Open your MPMB character sheet in Acrobat Reader DC or similar.
2. Enable the **Show extra features** tool bar, and click the Import/Export button
3. Pick the **Import/Export using files>Export .xfdf file>Export .xfdf file of all fields** option
4. Wait between 15s and several minutes depending on the complexity of your sheet. A new dialog will appear with XFDF output.
5. Select all the output in the window, and copy-and-paste and save into a new file with the XFDF extension. **WARNING**: The default text selected is not all of the file you need; make sure to `Ctrl-A, Ctrl-C` (or the equivalent on your platform) to get all the text
### On Foundry
6. On Foundry, select the Basic Controls on the left (the top-left "person" option, under the d20-anvil icon)
7. In that Basic Controls palette, click the **Import a PC character sheet** control (the person-plus option)
![The Import a PC option](https://github.com/spetzel2020/pc-importer/blob/master/img/import-pc-control.PNG?raw=true)
8. When the dialog opens, click the **Choose File** button, and browse to the location of the XFDF file created in Step 3 above and select that
9. Click **Import**
10. PC Importer will import recognized content into a newly created Actor (type character) and add unmatched content for the remaining classes, class features, feats, inventory, and spells. The name will show "...LOADING" after the Actor has been created, but while remaining content is being processed.
11. Open the character sheet for your new Actor; click the magnifying-glass icon to show matching options - you can right-click a match to either Add or Replace it in the sheet

## NEW: Manual matcher with Compendiums
When the import finishes, open the new character sheet and click on the (+) icon next to unmatched entries to match manually with Compendiums

## Contribution
[![ko-fi](https://www.ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/T6T82XFQD)

## License
**PC Importer for Foundry** by Jeffrey Pugh is licensed under a [GNU General Public License v3.0](https://github.com/spetzel2020/pc-importer/edit/master/LICENSE.md)

Portions incorporated from [Favourites Tab](https://github.com/syl3r86/favtab/blob/master/README.md)

This work is licensed under Foundry Virtual Tabletop [EULA - Limited License Agreement for module development v 0.1.6](http://foundryvtt.com/pages/license.html).


CHANGELOG.md
# Changelog
## 0.8.0
- Fixed: Issue [#3](https://github.com/spetzel2020/pc-importer/issues/3)

## 0.6.8
- Versioning and building artifacts

## 0.6.6
- Pick up Class Features in Extra.Notes

## 0.6.5
- Settings configuration sub-menu for choosing which Compendia are used
- Right-click option in Spell/Feature/Item matcher in Actor sheet: choose to Add or Replace


## 0.6.3
- Added Quantity and Weight to items
- Item matcher: use (+) icon to pop up comparison sheet; if you pick one it will be added to the Actor sheet
    (will need to add a better workflow)

- Speed up Foundry matching of spells, classes, items

## 0.6.2
 - Change import button to Person+
 - i18n button names on import dialog
 - Add ability to specify additional compendiums or re-order them (although you have to know their filenames)
 - Add race compendium and match Racial Features
 - Show LOADING on Actor name until it's done importing and matching	
- Fix up languages; move those not found in DND5E.languages into custom
- Default added weapons to Proficient

## 0.6.1
- MPMB: Correctly handle multi-classing
- MPMB: Item importing: match if you can, or import the name into the Loot category
        - changed to batch Item update
        - improved matching to find best match by intersecting words
- MPMB: Add Class Features, Feats
-   Tune matching with tradegoods and items so you don't get duplicates;
-   Tighten up matching based on excess words in either target or comparison


## 0.6.0
- Prototype importing from Fantasy Grounds

## 0.5.2
- Prototype Matching in Foundry on Class and Spell names

## 0.5.1
- Extract Spell and Class information as plugs to be later matched in Foundry

## 0.5.0a
- Changed repo and module names to pc-importer with the expectation that we might support importing PCs from other sources,
    e.g. MPMB, Fantasy Grounds, Roll 20, D&D Beyond
    Other importers exist, but this would provide a quick way of bringing over JUST a Player Character

## 0.4.3a
- Fill out inline functions to handle special case conversions: skills, HD, Proficiency, Senses, Personality_Trait, Spell slots
    (however some of these are ignored until we have loaded a class)
- Create Actor and then import JSON


## 0.4.2 (1-Oct-Sep-2020)
- More sophisticated Mapping template which allows for arrays, and uses null as the default
    - uses inline functions to define what should be done after mapping

## 0.4.1 (30-Sep-2020)
- Add more field mappings
- Try deconstructing xmlDoc instead of traversing the XML

## 0.4.0 (29-Sep-2020)
- Basic version of Actor5e.js to create an appropriate structure

## 0.3.0 (27-Sep-2020)
- Fix traversal and create flat (fieldName, value) dictionary

## 0.2.0 (27-Sep-2020)
- Read XML and traverse imperfectly

## 0.1.0 (25-Sep-2020)
- **Created**
-   Open an XFDF file from your local computer


modules/PCISettings.js
/*
29-Oct-2020     Created
30-Oct-2020     v0.6.4  Filter the presented compendiums to just those for the Item entity (includes spells, classes etc)
31-Oct-2020     v0.6.5b Add a submenu and a displayed (not editable) field for the list of compendiums
                Remove displayed field (can't keep it in sync in the static config form)
                Create subclasses of the submenu for each different Item type


*/
import {MODULE_NAME, MODULE_VERSION} from "./PCImporter.js";
import {defaultItemTypeToPackNames} from "./Actor5eFromExt.js";
import {buildItemPackIndexesByType} from "./MatchItem.js";




class PCISettings extends FormApplication {
    constructor(itemType, options={}) {
        super(itemType, options);
    }

    /** @override */
    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            title: game.i18n.localize("PCI.Compendia.SubMenu.LABEL"),
            id: "pc-importer",
            template: "modules/pc-importer/templates/compendium-select.html",
            top: 300,
            right: 300,
            width: 600,
            height: "auto",
            submitOnClose : true
            /*,
            tabs: [
                {navSelector: ".tabs", contentSelector: ".content", initial: "core"}
            ]
            */
        })
    }

    static init() {
        game.settings.register(MODULE_NAME, "PCImporterVersion", {
            name: "PC Importer version",
            hint: "",
            scope: "system",
            config: false,
            default: MODULE_VERSION,
            type: String
        });
        game.settings.register(MODULE_NAME, "loot", {
            name: "PCI.Compendia.Items.Setting.NAME",
            hint: "PCI.Compendia.Items.Setting.HINT",
            scope: "world",
            config: false,      //Derived value - don't show this
            default: defaultItemTypeToPackNames["loot"],
            type: String
        });
        game.settings.register(MODULE_NAME, "spell", {
            name: "PCI.Compendia.Spells.Setting.NAME",
            hint: "PCI.Compendia.Spells.Setting.HINT",
            scope: "world",
            config: false,      //Derived value - don't show this
            default: defaultItemTypeToPackNames["spell"],
            type: String
        });
        game.settings.register(MODULE_NAME, "feat", {
            name: "PCI.Compendia.Features.Setting.NAME",
            hint: "PCI.Compendia.Features.Setting.HINT",
            scope: "world",
            config: false,      //Derived value - don't show this
            default: defaultItemTypeToPackNames["feat"],
            type: String
        });
        game.settings.register(MODULE_NAME, "class", {
            name: "PCI.Compendia.Classes.Setting.NAME",
            hint: "PCI.Compendia.Classes.Setting.HINT",
            scope: "world",
            config: false,      //Derived value - don't show this
            default: defaultItemTypeToPackNames["class"],
            type: String
        });
        game.settings.registerMenu(MODULE_NAME, "classCompendiumsSubMenu", {
            name: "PCI.Compendia.Classes.Setting.NAME",
            label: "PCI.Compendia.SubMenu.LABEL",
            hint: "PCI.Compendia.SubMenu.HINT",
            icon: "fas fa-th-list",
            type: PCISettingsClass,
            restricted: true
        });
        game.settings.registerMenu(MODULE_NAME, "featCompendiumsSubMenu", {
            name: "PCI.Compendia.Features.Setting.NAME",
            label: "PCI.Compendia.SubMenu.LABEL",
            hint: "PCI.Compendia.SubMenu.HINT",
            icon: "fas fa-th-list",
            type: PCISettingsFeature,
            restricted: true
        });
        game.settings.registerMenu(MODULE_NAME, "lootCompendiumsSubMenu", {
            name: "PCI.Compendia.Items.Setting.NAME",
            label: "PCI.Compendia.SubMenu.LABEL",
            hint: "PCI.Compendia.SubMenu.HINT",
            icon: "fas fa-th-list",
            type: PCISettingsItem,
            restricted: true
        });
        game.settings.registerMenu(MODULE_NAME, "spellCompendiumsSubMenu", {
            name: "PCI.Compendia.Spells.Setting.NAME",
            label: "PCI.Compendia.SubMenu.LABEL",
            hint: "PCI.Compendia.SubMenu.HINT",
            icon: "fas fa-th-list",
            type: PCISettingsSpell,
            restricted: true
        });

    }

    /** @override */
    async getData() {
        const entity = "Item";
        const currentlySelectedCompendiums = game.settings.get(MODULE_NAME, this.object);
        const currentCompendiumsArray = currentlySelectedCompendiums.split(",");
        //Convert game.packs from Map to Array
        const packsForEntity = game.packs.filter(v => v.metadata.entity === entity);
        const compendiumsWithSelectionFlag = packsForEntity.map(p => {
            const name = p.collection;
            const isSelected = currentCompendiumsArray.includes(name);
            return {compendium : p, isSelected}
        });
        return {
            itemType: this.object,
            compendiums: compendiumsWithSelectionFlag,
            listLength: compendiumsWithSelectionFlag.length,
            default: defaultItemTypeToPackNames[this.object],
            selectedCompendiums : currentlySelectedCompendiums
        };
    }
    
    async _updateObject(event, formData) {
        //Update the hidden values of compendiums
        //Use formData because a Close does not set the form value properly
        const selectedCompendiums = formData["selected-compendiums"];
        await game.settings.set(MODULE_NAME, this.object, selectedCompendiums);
        //Rebuild the indexes for the item matching in the actor form
        buildItemPackIndexesByType();
    }

    _onChangeInput(event) {
        //Update the hidden values of compendiums
        const selectedCompendiums = $("#select-compendiums").val();
        const concatCompendiums = selectedCompendiums.join();
        //Set the text in the "selected-compendiums" field
        let concatField = document.getElementById("selected-compendiums");
        if (concatField) concatField.value = concatCompendiums;

    }


}

class PCISettingsClass extends PCISettings {
    constructor() {
        super("class");
    }
}
class PCISettingsItem extends PCISettings {
    constructor() {
        super("loot");
    }
}
class PCISettingsSpell extends PCISettings {
    constructor() {
        super("spell");
    }
}
class PCISettingsFeature extends PCISettings {
    constructor() {
        super("feat");
    }
}

Hooks.on("init", PCISettings.init);

modules/Actor5e.js
/*
29-Sep-2020     Created
30-Sep-2020     Mapping template: provide the key and the default if it's not found (because some defaults are empty arrays etc)
1-Oct-2020      v0.4.2: Embed functions in the mapping template to guide specific calculations
                basically replaces inline code in an importer/constructor
                Handle: lowercase languages, flags which are True/Off -> 1/0, trim " " to null
                v0.4.3: Add calculations/functions for Skills. Here value = 0,1, or 2 depending on Not Prof/Prof/Expertise
                Not implmenting any bonuses or other adjustments, and we're hoping the totals will recalculate automatically
2-Oct-2020      v0.4.3a: createFoundryActor(): Imports the JSON just created.


TO DO:
    - Import the basic JSON into the importer and create the new Actor
    - THEN, import the Items consisting of Class(es), Inventory etc, either by referencing SRD or just copying directly
        - suggests a different format for the Item values - not perhaps embedded in the Actor
    - Do a Regex match on the Spells and Items
    - Perhaps use Object.create to create the skeleton Actor5eFromMPMB and then populate it from the fieldDictionary the other way round
        using dot or array notation
    - Handle multi-classed characters
    - Import the JSON directly in and then make addition adjustments for Classes

*/

import {MODULE_NAME, MPMBImporter} from "./MPMBImporter.js";
import Actor5e from "/systems/dnd5e/module/actor/entity.js";    //default export

const Ability = {
    str : "str",
    dex : "dex",
    con : "con",
    int : "int",
    wis : "wis",
    cha : "cha"
}


export class Actor5eFromMPMB {
    constructor(importedDictionary) {
        this.fieldDictionary = importedDictionary;
        this.actorJSON = null;
        //IMPORTING

        //See https://stackoverflow.com/questions/8085004/iterate-through-nested-javascript-objects
        //Use recursion (only 3 levels deep) to walk through the mapping tree and object in parallel
        this.mapAndIterate(MPMBtoActor5eMapping, this);
        delete this.fieldDictionary;
    }

    mapAndIterate(subMappingTree, subObject) {
        for (const entry of Object.entries(subMappingTree)) {
            const entryProperty = entry[0];
            const entryValue = entry[1];
            //If itself an object, then keep on drilling down
            //0.4.2 Have to run it through the mapper because we have both legitimate nested objects
            //and also the special mapping objects like {fieldName: , default: }
            let mappedValue = this.mapToMPMBField(entryValue);
            subObject[entryProperty] = mappedValue;
        }
        return subObject;
    }

    mapToMPMBField(entryValue) {
        //v0.4.2 Allow for different forms of the mapping variable
        //  null -> do nothing
        //  {fieldName, default} -> look up the fieldName (if present); use the default if not found
        //  String -> look up the fieldName and use null if not found
        // Array -> do all these things for each element of the array
        // Empty object - will probably become null
        let mappedValue;
        if (!entryValue) {
            mappedValue = mappedValue;  //Works if the default is null or 0
        } else if (Array.isArray(entryValue)) {
            //Not sure this will handle all possibilities
            //Should handle both the empty array [] and also [fieldName1, fieldName2, ....]
            const mappedArray = [];
            entryValue.forEach(elem => {
                const mappedArrayValue = this.mapToMPMBField(elem);
                if (mappedArrayValue) {mappedArray.push(mappedArrayValue);}
            });
            mappedValue = mappedArray;
        } else if (entryValue.fieldName) {
            mappedValue = MPMBImporter.getValueForFieldName(this.fieldDictionary, entryValue.fieldName);
            if (!mappedValue) {mappedValue = entryValue.default ? entryValue.default : null}
        } else if (typeof entryValue === 'function') {
            mappedValue = entryValue(this.fieldDictionary);
        } else if (entryValue.default) {
            mappedValue = entryValue.default;
        } else if (typeof entryValue === 'object') {
            //This is for the nested form of the mapping template
            const subObject = {}
            mappedValue = this.mapAndIterate(entryValue, subObject);
        } else {
            //Trims spaces from beginning and end (so turns " " into "")
            mappedValue = MPMBImporter.getValueForFieldName(this.fieldDictionary, entryValue);
            if (typeof mappedValue === 'string') {mappedValue = mappedValue.trim();}
        }
        //FXIME: Do some clean-up - would be better if this could be done inline
        if (mappedValue ==="Off") {mappedValue = false;}
        if (mappedValue === "") {mappedValue = null;}
        return mappedValue;
    }

    exportToJSON(createFile=true) {
        let data = duplicate(this);
        let allData = null;

        // Flag some metadata about where the entity was exported some - in case migration is needed later
        //Redudant since we're storing this on every element
        //data.flags["exportSource"] = metadata;
        const dataAsJSON = JSON.stringify(data, null, 2);

        // Trigger file save procedure
        if (createFile) {
            const filename = `fvtt-${this.name}.json`;
            saveDataToFile(dataAsJSON, "text/json", filename);
        }
        this.actorJSON = dataAsJSON;
        return dataAsJSON;
    }

    async createFoundryActor() {
        if (!this.actorJSON) {return false;}
        let actorData = duplicate(this);
        delete actorData.actorJSON;
        const options = {temporary: false, renderSheet: false}
        const newActor = await Actor5e.create(actorData, options);
        await newActor.importFromJSON(this.actorJSON);
        newActor.initialize();
    }
}

function defaulted(defaultValue) {return defaultValue;}

function lowercaseArray(mappedArray) {
    mappedArray.forEach((elem,i) => {mappedArray[i] = elem.toLowerCase()});
    return mappedArray;
}

function mapArray(dictionary, fieldNames) {
    const fieldNameArray = Array.from(fieldNames);
    let mappedArray = [];
    fieldNameArray.forEach((fieldName, i) => {
        const mappedValue = MPMBImporter.getValueForFieldName(dictionary, fieldName);
        if (mappedValue && (mappedValue !== " ")) {mappedArray.push(mappedValue);}
    });
    return mappedArray;
}

function mapAndSwitchToInteger(dictionary,fieldName) {
    const switchValue = MPMBImporter.getValueForFieldName(dictionary, fieldName);
    if (switchValue === "Off") {return 0;}
    else if (switchValue === "True") {return 1;}
    else {return 0;}
}
function mapConvertAndAdd(dictionary,fieldNames) {
    //Used for turning Proficiency + Expertise -> 0, 1, or 2 (which is what Foundry wants)
    if (!fieldNames || !fieldNames.length || !Array.isArray(fieldNames)) {return 0;}
    const converted = fieldNames.map(k => MPMBImporter.getValueForFieldName(dictionary, k));
    const added = converted.reduce((sum,c) => sum + (c==="True" ? 1 : 0),0);
    return added;
}




//Provides the mapping to the field names in the XFDF
const MPMBtoActor5eMapping = {
      "name": "AdvLog.PC Name",
      "type": () => "character",
      "flags": {
        "exportSource": {
          "world": null,
          "system": {default: "dnd5e"},
          "coreVersion": {default: "0.6.6"},
          "systemVersion": {default: "0.96"}
        },
        "dnd5e": {},
        "core": {
          "sheetClass": {default: ""}
        },
        "MPMB Importer" : {default: "0.4.2"}
      },
      "data": {
        "abilities": {
          "str": {
            "value": "Str",
            "proficient":  function(dictionary,fieldName="Str ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Str Mod",
            "prof": 0,      //Don't need to set? Is calculated?
            "saveBonus": 0,     //Only if you have an item (e.g. +1 CLoak of Resistance)
            "checkBonus": 0,
            "save": "Str ST Mod"
          },
          "dex": {
            "value": "Dex",
            "proficient": function(dictionary,fieldName="Dex ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Dex Mod",
            "prof": 0,
            "saveBonus": 0,
            "checkBonus": 0,
            "save": "Dex ST Mod"
          },
          "con": {
            "value": "Con",
            "proficient":  function(dictionary,fieldName="Con ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Con Mod",
            "prof": 0,
            "saveBonus": 0,
            "checkBonus": 0,
            "save": "Con ST Mod"
          },
          "int": {
            "value": "Int",
            "proficient":  function(dictionary,fieldName="Int ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Int Mod",
            "prof": 0,
            "saveBonus": 0,
            "checkBonus": 0,
            "save": "Int ST Mod"
          },
          "wis": {
            "value": "Wis",
            "proficient":  function(dictionary,fieldName="Wis ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Wis Mod",
            "prof": 0,
            "saveBonus": 0,
            "checkBonus": 0,
            "save": "Wis ST Mod"
          },
          "cha": {
            "value": "Cha",
            "proficient": function(dictionary,fieldName="Cha ST Prof") {return mapAndSwitchToInteger(dictionary,fieldName);},
            "mod": "Cha Mod",
            "prof": 0,
            "saveBonus": 0,
            "checkBonus": 0,
            "save": "Cha ST Mod"
          }
        },
        "attributes": {
          "ac": {
            "value": "AC"
          },
          "hp": {
            "value": "HP Current",
            "min": 0,
            "max": "HP Max",
            "temp": "HP Temp",
            "tempmax": 0
          },
          "init": {
            "value": "Initiative Bonus",
            "bonus": 0,
            "mod": 0,
            "prof": 0,
            "total": 0
          },
          "spellcasting": "int",
          "speed": {
            "value": "Speed",
            "special": null
          },
          "death": {
            "success": null,
            "failure": null
          },
          "encumbrance": {
            "value": 0,
            "max": 150,
            "pct": 0,
            "encumbered": false
          },
          "exhaustion": null,
          "inspiration": "Inspiration",
          "hd": "HD1 Level",
          "prof": "Proficiency Bonus",
          "spelldc": "Spell save DC 1"
        },
        "details": {
          "biography": {
            "value": null,
            "public": "Background History"
          },
          "alignment": "Alignment",
          "race": "Race",
          "background": "Background",
          "xp": {
            "value": "Total Experience",
            "min": null,
            "max": null,
            "pct": null
          },
          "trait": "Personality Trait",
          "ideal": "Ideal",
          "bond": "Bond",
          "flaw": "Flaw",
          "level": "Character Level"
        },
        "traits": {
          "size": "Size Category",
          "di": {
            "value": [],
            "custom": null
          },
          "dr": {
            "value":  function(dictionary) {return mapArray(dictionary, ["Resistance Damage Type 1","Resistance Damage Type 2","Resistance Damage Type 3","Resistance Damage Type 4","Resistance Damage Type 5"]);},
            "custom": null
          },
          "dv": {
            "value": [],
            "custom": null
          },
          "ci": {
            "value": [],
            "custom": null
          },
          "senses": "Vision",
          "languages": {
            "value":  function(dictionary) {return lowercaseArray(mapArray(dictionary, ["Language 1","Language 2","Language 3","Language 4","Language 5"]));},
            "custom": null
          },
          "weaponProf": {
            "value": [],
            "custom": null
          },
          "armorProf": {
            "value": [],
            "custom": null
          },
          "toolProf": {
            "value": [],
            "custom": null
          }
        },
        "currency": {
          "pp": "Platinum Pieces",
          "gp": "Gold Pieces",
          "ep": "Electrum Pieces",
          "sp": "Silver Pieces",
          "cp": "Copper Pieces"
        },
        "skills": { //"bonus" is global bonus, for example from item, magic or otherwise
          "acr": {  //Acrobatics
            "value":  function(dictionary,fieldNames=["Acr Exp","Acr Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "dex",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "ani": {  //Animal handling
            "value": function(dictionary,fieldNames=["Ani Exp","Ani Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "wis",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "arc": {  //Arcana
            "value": function(dictionary,fieldNames=["Arc Exp","Arc Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "int",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "ath": {  //Athletics
            "value": function(dictionary,fieldNames=["Ath Exp","Ath Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "str",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "dec": {  //Deception
            "value": function(dictionary,fieldNames=["Dec Exp","Dec Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "cha",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "his": {  //History
            "value": function(dictionary,fieldNames=["His Exp","His Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "int",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "ins": {  //Insigh
            "value": function(dictionary,fieldNames=["Ins Exp","Ins Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "wis",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "itm": {
            "value": function(dictionary,fieldNames=["Acr Exp","Acr Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "cha",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "inv": {  //Investigation
            "value": function(dictionary,fieldNames=["Inv Exp","Inv Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "int",
            "bonus": "Inv Bonus",
            "mod":  null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "med": {  //Medicine
            "value": function(dictionary,fieldNames=["Med Exp","Med Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "wis",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "nat": {  //Nature
            "value": function(dictionary,fieldNames=["Nat Exp","Nat Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "int",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "prc": {  //Perception
            "value": function(dictionary,fieldNames=["Perc Exp","Perc Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "wis",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": "Passive Perception"
          },
          "prf": {  //Performance
            "value": function(dictionary,fieldNames=["Perf Exp","Perf Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "cha",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "per": {  //Persuasion
            "value": function(dictionary,fieldNames=["Pers Exp","Pers Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "cha",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "rel": {  //Religion
            "value": function(dictionary,fieldNames=["Rel Exp","Rel Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "int",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "slt": {  //Sleight of Hand
            "value": function(dictionary,fieldNames=["Sle Exp","Sle Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "dex",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "ste": {  //Stealth
            "value": function(dictionary,fieldNames=["Ste Exp","Ste Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "dex",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          },
          "sur": {  //Survival
            "value": function(dictionary,fieldNames=["Sur Exp","Sur Prof"]) {return mapConvertAndAdd(dictionary,fieldNames);},
            "ability": "wis",
            "bonus": null,
            "mod": null,
            "prof": null,
            "total": null,
            "passive": null
          }
        },
        "spells": {
          "spell1": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl1",
            "override": null
          },
          "spell2": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl2",
            "override": null
          },
          "spell3": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl3",
            "override": null
          },
          "spell4": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl4",
            "override": null
          },
          "spell5": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl5",
            "override": null
          },
          "spell6": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl6",
            "override": null
          },
          "spell7": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl7",
            "override": null
          },
          "spell8": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl8",
            "override": null
          },
          "spell9": {
            "value": null,
            "max": "SpellSlots.CheckboxesSet.lvl9",
            "override": null
          },
          "pact": {
            "value": null,
            "max": null,
            "override": null,
            "level": null
          }
        },
        "bonuses": {
          "mwak": {
            "attack": null,
            "damage": null
          },
          "rwak": {
            "attack": null,
            "damage": null
          },
          "msak": {
            "attack": null,
            "damage": null
          },
          "rsak": {
            "attack": null,
            "damage": null
          },
          "abilities": {
            "check": null,
            "save": null,
            "skill": null
          },
          "spell": {
            "dc": null
          }
        },
        "resources": {
          "primary": {
            "value": null,
            "max": null,
            "sr": false,
            "lr": false,
            "label": null
          },
          "secondary": {
            "value": null,
            "max": null,
            "sr": false,
            "lr": false,
            "label": null
          },
          "tertiary": {
            "value": null,
            "max": null,
            "sr": false,
            "lr": false,
            "label": null
          }
        }
      },
      "sort": null,

      "token": {
        "flags": {},
        "name": null,
        "displayName": null,
        "img": "icons/svg/mystery-man.svg",
        "tint": null,
        "width": 1,
        "height": 1,
        "scale": 1,
        "lockRotation": false,
        "rotation": null,
        "vision": true,
        "dimSight": 60,
        "brightSight": null,
        "dimLight": null,
        "brightLight": null,
        "sightAngle": 360,
        "lightAngle": 360,
        "lightAlpha": 1,
        "actorId": null,
        "actorLink": true,
        "actorData": null,
        "disposition": 1,
        "displayBars": null,
        "bar1": {
          "attribute": {default: "attributes.hp"}
        },
        "bar2": {
          "attribute": ""
        },
        "randomImg": false
    },
    "_id": null,
    "img": null
  }
 const MPMBToItemMapping = {
     //We add multiple of these to the created Actor
      "item": {   //Foundry uses this for lots of stuff, including Class levels
            "name": {default: "Wizard"},
            "type": {default: "class"},
            "img": "systems/dnd5e/icons/skills/blue_13.jpg",
            "data": {
              "description": {
                "value": "",
                "chat": "",
                "unidentified": ""
              },
              "source": "",
              "levels": "Character Level",
              "subclass": "Class and Levels",
              "hitDice": "HD1 Dice",
              "hitDiceUsed": 0,
              "skills":{},
              "spellcasting": "full",
              "damage": {
                "parts": []
              }
            }
      }
}
